src/search/domain_abstractions/domain_abstraction.cc

```diff
@@ -0,0 +1,365 @@
+            transition_system(std::move(transitionSystem)),
+            concrete_initial_state(originalTask.get_initial_state()),
+            goal_facts(task_properties::get_fact_pairs(originalTask.get_goals())),
+            originalTask(originalTask),
+            log(log),
+            variableGroupVectors(std::move(domainMap)),
+            nValuesForHash(variableGroupVectors.size(), 0),
+            operatorCosts(task_properties::get_operator_costs(originalTask)),
+            domainSizes(variableGroupVectors.size(), 1) {
+        // precompute N-Values for perfect hash function that is needed for lookup
+        reload(variableGroupVectors);
+    }
+
+    vector<long long>
+    DomainAbstraction::computeNValues(VariableGroupVectors newAbstraction, vector<int> &newDomainSizes) {
+        /*
+         * uses current AbstractDomains to calculate the N Values that are needed for the perfect hash function.
+         * The perfect hash function maps Abstract states to an index which is used for lookup h values and to
+         * store them when computed
+         * */
+        // loop over all variables (we always have all variables considered in comparison to this hash for PDB)
+        vector<long long> newNValuesForHash(newAbstraction.size(), 1);
+        for (int i = 0; i < (int) newAbstraction.size(); i++) {
+            // loop over groups in abstract domain for current variable v_i
+            for (int j = 0; j <= i - 1; j++) {
+                VariableGroupVector varGroupMapping = newAbstraction[j];
+                long long numGroups = newDomainSizes.at(j);
+                assert(numGroups > 0);
+                newNValuesForHash.at(i) *= numGroups;
+            }
+            assert(newNValuesForHash.at(i) >= 0);
+        }
+        assert(nValuesForHash.size() == variableGroupVectors.size());
+        return newNValuesForHash;
+    }
+
+    bool DomainAbstraction::isGoal(const shared_ptr <DomainAbstractedState> &candidate) {
+        /*
+         * Check if the abstract state id is present in goalAbstractStates List -> if yes it is a goal state
+         * */
+        vector<int> candidateGroupAssignments = candidate->getGroupsAssignment();
+        return groupAssignmentFulfillsFacts(candidateGroupAssignments, goal_facts);
+    }
+
+
+    DomainAbstractedStates DomainAbstraction::getSuccessors(const shared_ptr <DomainAbstractedState> &state) {
+        /*
+         * Returns a List of Abstract states (DomainAbstractedState) (representation of one state = vec<int>) that are the
+         * possible successors of the given Abstract state. It uses a map to store already found successors, so that duplicated are
+         * */
+        map<long long, shared_ptr<DomainAbstractedState>> alreadyFound; // store Id's of already found successors
+
+        // loop over all operators
+        for (int operatorIndex = 0; operatorIndex < transition_system->get_num_operators(); operatorIndex++) {
```

Comment:
It might be a useful optimization to store for each abstract state which operators are applicable. Doing so would speed up the search time at the cost of slowing down the precomputation. Since our bottleneck is in the search and not in the precomputation, this could pay off. I would leave this as an idea to persue if we have time left in the end.

Reply:
This could actually also speedup the precomputation I think. This is a very good Idea, as I already compute the abstract post and preconditions in advance to the precomputation this is not very difficult to do.